File: MtxVec.Delphi/FileHandling1.pas
unit FileHandling1;

interface
uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Dialogs,
  Basic2, StdCtrls, ExtCtrls, ActnList, MtxVec,MtxVecEdit, ComCtrls,
  Buttons, MtxExpr;

type
  TFileH1 = class(TBasicForm2)
    CheckBox1: TCheckBox;
    Label1: TLabel;
    RichEdit2: TRichEdit;
    Button1: TBitBtn;                            
    Button2: TBitBtn;
    Button3: TBitBtn;
    Button4: TBitBtn;
    procedure FormCreate(Sender: TObject);
    procedure Button1Click(Sender: TObject);
    procedure Button2Click(Sender: TObject);
    procedure Button4Click(Sender: TObject);
    procedure Button3Click(Sender: TObject);
  private
    AMtx: Matrix;
    AVec: Vector;
    { Private declarations }
  public
    { Public declarations }
  end;

var
  FileH1: TFileH1;

implementation

{$IFDEF CLR}
{$R *.NFM}
{$ELSE}
{$R *.DFM}
{$ENDIF}

procedure TFileH1.FormCreate(Sender: TObject);
begin
  inherited;
  With RichEdit1.Lines do
  begin
    Clear;
    Add('Importing and exporting values is an important feature '
      + 'of MtxVec. MtxVec offers you the ability to save vector/'
      + 'matrix to memory stream, binary files, text files, BLOBs, Clipboard ...');
    Add('Click on buttons bellow to introduce yourself with various '
      + 'methods available. If "View loaded values" option is '
      + 'checked, you''ll get a visual representation of loaded matrix/vector.');
  end;
end;

procedure TFileH1.Button1Click(Sender: TObject);
var AStream : TMemoryStream;
    tmpMtx : Matrix;
{    AStream1: MemoryStream; }
begin
  With RichEdit2.Lines, RichEdit2 do
  begin
    Clear;
    Add('AMtx.Size(10,10);');
    Add('AMtx.RandGauss; {add random data }');
    SelAttributes.Style := [fsItalic];
    Add('{ save header & values to stream}');
    Add('AStream := TMemoryStream.Create;');
    Add('try');
    Add(' AMtx.SaveToStream(AStream);');
    SelAttributes.Style := [fsItalic];
    Add(' { now load header & values from stream }');
    Add(' CreateIt(tmpMtx);');
    Add(' try');
    Add('   AStream.Seek(0,0);');
    Add('   tmpMtx.LoadFromStream(AStream);');
    Add('   If CheckBox1.Checked then ViewValues(AMtx);');
    Add(' finally');
    Add('   FreeIt(tmpMtx);');
    Add(' end;');
    Add('finally');
    Add('  AStream.Free;');
    Add('end;');
  end;
  AMtx.Size(10,10);
  AMtx.RandGauss; {add random data }

  {$IFNDEF CLR}

  { save header & values to stream}
  AStream := TMemoryStream.Create;
  try
     AMtx.SaveToStream(AStream);
    { now load header & values from stream }
     {$IFDEF CLR}
     AStream.Seek(0,soFromBeginning);
     {$ELSE}
     AStream.Seek(0,0);
     {$ENDIF}
     tmpMtx.LoadFromStream(AStream);
     If CheckBox1.Checked then ViewValues(AMtx,'Memory stream',true);
  finally
     AStream.Free;
  end;

  {$ELSE}

  { save header & values to stream}
  AStream1 := MemoryStream.Create;
  try
     AMtx.SaveToStream(AStream1);
    { now load header & values from stream }
     {$IFDEF CLR}
     AStream1.Seek(0,SeekOrigin.Begin);
     {$ELSE}
     AStream1.Seek(0,0);
     {$ENDIF}
     tmpMtx.LoadFromStream(AStream1);
     If CheckBox1.Checked then ViewValues(AMtx,'Memory stream',true);
  finally
     AStream.Free;
  end;

  {$ENDIF}
end;

procedure TFileH1.Button2Click(Sender: TObject);
var tmpVec: Vector;
begin
  With RichEdit2.Lines, RichEdit2 do
  begin
    Clear;
    Add('AVec.Size(100);');
    Add('AVec.RandUniform(-2,2);');
    SelAttributes.Style := [fsItalic];
    Add('{save header & values to file}');
    Add('AVec.SaveToFile(''TestVec1.vec'');');
    Add('CreateIt(tmpVec);');
    Add('try');
    SelAttributes.Style := [fsItalic];
    Add(' { load header & values into tmpVec}');
    Add(' tmpVec.LoadFromFile(''TestVec1.vec'');');
    Add(' if CheckBox1.Checked then ViewValues(tmpVec);');
    Add('finally');
    Add(' FreeIt(tmpVec);');
    Add('end;');
  end;
  AVec.Size(100);
  AVec.RandUniform(-2,2);
  {save header & values to file}
  AVec.SaveToFile('TestVec1.vec');

  { load header & values into tmpVec}
  tmpVec.LoadFromFile('TestVec1.vec');
  if CheckBox1.Checked then ViewValues(tmpVec,'File Stream',true);
end;

procedure TFileH1.Button4Click(Sender: TObject);
var StringList : TStringList;
    tmpMtx: Matrix;
begin

  With RichEdit2.Lines, RichEdit2 do
  begin
    Clear;
    Add('AMtx.Size(20,20,true);');
    Add('AMtx.RandUniform(-1,2);');
    Add('StringList := TStringList.Create;');
    Add('try');
    SelAttributes.Style := [fsItalic];
    Add(' { use tab = chr(9) as delimiter }');
    Add(' AMtx.ValuesToStrings(StringList,#9);');
    SelAttributes.Style := [fsItalic];
    Add(' { Save matrix values to txt file }');
    Add(' StringList.SaveToFile(''ASCIIMtx.txt'');');
    Add('finally');
    Add(' StringList.Free;');
    Add('end;');
    Add('');
    Add('StringList := TStringList.Create;');
    Add('CreateIt(tmpMtx);');
    Add('try');
    SelAttributes.Style := [fsItalic];
    Add(' { get matrix values from text file }');
    Add(' StringList.LoadFromFile(''ASCIIMtx.txt'');');
    SelAttributes.Style := [fsItalic];
    Add(' { use tab = chr(9) as delimiter }');
    Add(' tmpMtx.StringsToValues(StringList,#9);');
    Add(' if CheckBox1.Checked then ViewValues(tmpMtx);');
    Add('finally');
    Add(' FreeIt(tmpMtx);');
    Add(' StringList.Free;');
    AdD('end;');
  end;

  AMtx.Size(20,20,true);
  AMtx.RandUniform(-1,2);
  StringList := TStringList.Create;
  try
     { use tab = chr(9) as delimiter }
     AMtx.ValuesToStrings(StringList,#9);
     { Save matrix values to txt file }
     StringList.SaveToFile('ASCIIMtx.txt');
  finally
     StringList.Free;
  end;

  StringList := TStringList.Create;
  try
     { get matrix values from text file }
     StringList.LoadFromFile('ASCIIMtx.txt');
     { use tab = chr(9) as delimiter }
     tmpMtx.StringsToValues(StringList,#9);
     if CheckBox1.Checked then ViewValues(tmpMtx,'TEXT file',true);
  finally
     StringList.Free;
  end;
end;

procedure TFileH1.Button3Click(Sender: TObject);
(*
var
    tmpBlob : TStream;
    Table1  : TTable;
*)
begin
  With RichEdit2.Lines, RichEdit2 do
  begin
    Clear;
    Add('AVec.Size(100);');
    Add('AVec.RandGauss(0,1);');
    Add('AVec.SortAscend;');
    Add('Table1 := TTable.Create(Self);');
    Add('With Table1 do');
    Add(' try');
    Add(' DatabaseName := ''DBDemos'';');
    Add(' TableType := ttDefault;');
    Add(' TableName := ''TableVec'';');
    Add(' with FieldDefs do');
    Add(' begin');
    Add('   Clear;');
    Add('   with AddFieldDef do');
    Add('   begin');
    Add('     Name := ''Field1'';');
    Add('     DataType := ftBLOB;');
    Add('     Required := False;');
    Add('   end;');
    Add(' end;');
    Add(' CreateTable;');
    SelAttributes.Style := [fsItalic];
    Add(' { save to table1}');
    Add(' Active := true;');
    Add(' Insert;');
    Add(' tmpBLOB := CreateBLOBStream(FieldByName(''Field1''),bmReadWrite);');
    Add(' CreateIt(tmpVec);');
    Add(' try');
    Add('   AVec.SaveToStream(tmpBlob);');
    Add('   Post;');
    Add('   tmpBlob.Seek(0,0); { "reset" BLOB stream}');
    Add('   tmpVec.LoadFromStream(tmpBlob);');
    Add('   If CheckBox1.Checked then ViewValues(tmpVec);');
    Add(' finally');
    Add('   FreeIt(tmpVec);');
    Add('   tmpBLOB.Free;');
    Add(' end;');
    Add('finally');
    Add(' Active := false;');
    Add(' DeleteTable;');
    Add(' Destroy;');
    Add('end;');
  end;
{
  AVec.Size(100);
  AVec.RandGauss(0,1);
  AVec.SortAscend;
  Table1 := TTable.Create(Self);
  With Table1 do
  try
     DatabaseName := 'DBDemos';
     TableType := ttDefault;
     TableName :='TableVec';
     with FieldDefs do
     begin
      Clear;
      with AddFieldDef do
      begin
        Name := 'Field1';
        DataType := ftBlob;
        Required := False;
      end;
     end;
     CreateTable;

  //   save to table1
     Active := true;
     Insert;
     tmpBLOB := CreateBLOBStream(FieldByName('Field1'),bmReadWrite);
     CreateIt(tmpVec);
     try
        AVec.SaveToStream(tmpBlob);
        Post;
        tmpBlob.Seek(0,0); // "reset" BLOB stream
        tmpVec.LoadFromStream(tmpBlob);
        if CheckBox1.Checked then ViewValues(tmpVec,'BLOB Stream',true);
     finally
        FreeIt(tmpVec);
        tmpBLOB.Free;
     end;
  finally
     Active := false;
     DeleteTable;
     Destroy;
  end;
}
end;

initialization
   RegisterClass(TFileH1);
end.

File: MtxVec.CPP/FileHandling1.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

// uncomment if BDE is used
// #define BDE

#ifdef BDE
  #include <Db.hpp>
  #include <DBTables.hpp>
#endif


#include "FileHandling1.h"
#include "MtxVecEdit.hpp"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "Basic2"
#pragma resource "*.dfm"
TFileH1 *FileH1;
//---------------------------------------------------------------------------
__fastcall TFileH1::TFileH1(TComponent* Owner)
  : TBasicForm2(Owner)
{
  AnsiString str;

  RichEdit1->Lines->Clear();
  str = "Importing and exporting values is an important feature ";
  str+= "of MtxVec. MtxVec offers you the ability to save vector/";
  str+= "matrix to memory stream, binary files, text files, BLOBs, Clipboard ...";
  RichEdit1->Lines->Add(str);
  str = "Click on buttons bellow to introduce yourself with various ";
  str+= "methods available. If ""View loaded values"" option is ";
  str+= "checked, you'll get a visual representation of loaded matrix/vector.";
  RichEdit1->Lines->Add(str);
}
//---------------------------------------------------------------------------


void __fastcall TFileH1::Button1Click(TObject *Sender)
{
  sMatrix tmpMtx;
  RichEdit2->Lines->Clear();

  RichEdit2->Lines->Add("AMtx->Size(10,10);");
  RichEdit2->Lines->Add("AMtx->RandGauss(); //add random data }");
  RichEdit2->SelAttributes->Style = RichEdit2->SelAttributes->Style << TFontStyle::fsItalic;
  RichEdit2->Lines->Add("// save header & values to stream}");
  RichEdit2->Lines->Add("TMemoryStream AStream = TMemoryStream();");
  RichEdit2->Lines->Add("try {");
  RichEdit2->Lines->Add(" AMtx->SaveToStream(AStream);");
  RichEdit2->SelAttributes->Style = RichEdit2->SelAttributes->Style << TFontStyle::fsItalic;
  RichEdit2->Lines->Add(" // now load header & values from stream }");
  RichEdit2->Lines->Add("   AStream->Seek(0,0);");
  RichEdit2->Lines->Add("   tmpMtx->LoadFromStream(AStream);");
  RichEdit2->Lines->Add("   If (CheckBox1->Checked) ViewValues(AMtx);");
  RichEdit2->Lines->Add("} __finally {");
  RichEdit2->Lines->Add("");
  RichEdit2->Lines->Add("  delete AStream");
  RichEdit2->Lines->Add("");
  RichEdit2->Lines->Add("}");


  AMtx.Size(10,10,false);
  AMtx.RandGauss(); // add random data

  // save header & values to stream}
  TMemoryStream *AStream = new TMemoryStream;
  try
  {
	AMtx.SaveToStream(AStream);
	// now load header & values from stream
	AStream->Seek(0,0);
	tmpMtx.LoadFromStream(AStream);
	if (CheckBox1->Checked) ViewValues(AMtx,"Memory stream",true,false,true);
  }
  __finally
  {
    delete AStream;
  }
}
//---------------------------------------------------------------------------

void __fastcall TFileH1::Button2Click(TObject *Sender)
{
  sVector tmpVec;
  RichEdit2->Lines->Clear();

  RichEdit2->Lines->Add("AVec->Size(100);");
  RichEdit2->Lines->Add("AVec.RandUniform(-2,2);");
  RichEdit2->SelAttributes->Style = RichEdit2->SelAttributes->Style << TFontStyle::fsItalic;
  RichEdit2->Lines->Add(" //save header & values to file}");
  RichEdit2->Lines->Add("AVec->SaveToFile(""TestVec1.vec"");");
  RichEdit2->SelAttributes->Style = RichEdit2->SelAttributes->Style << TFontStyle::fsItalic;
  RichEdit2->Lines->Add(" // load header & values into tmpVec}");
  RichEdit2->Lines->Add("tmpVec->LoadFromFile(""TestVec1.vec"");");
  RichEdit2->Lines->Add("if (CheckBox1->Checked) ViewValues(tmpVec);");

  AVec.Size(100);
  AVec.RandUniform(-2,2);
  AVec.SaveToFile("TestVec1.vec");   // save header & values to file

  tmpVec.LoadFromFile("TestVec1.vec");     // load header & values into tmpVec
  if (CheckBox1->Checked) ViewValues(tmpVec,"File Stream",true);
}
//---------------------------------------------------------------------------

void __fastcall TFileH1::Button4Click(TObject *Sender)
{
  sMatrix tmpMtx;
  RichEdit2->Lines->Clear();

  RichEdit2->Lines->Add("AMtx->Size(20,20,true);");
  RichEdit2->Lines->Add("AMtx->RandUniform(-1,2);");
  RichEdit2->Lines->Add("TStringList StringList = new TStringList();");
  RichEdit2->Lines->Add("try {");
  RichEdit2->SelAttributes->Style = RichEdit2->SelAttributes->Style << TFontStyle::fsItalic;
  RichEdit2->Lines->Add(" // use tab = chr(9) as delimiter ");
  RichEdit2->Lines->Add(" AMtx->ValuesToStrings(StringList,#9);");
  RichEdit2->SelAttributes->Style = RichEdit2->SelAttributes->Style << TFontStyle::fsItalic;
  RichEdit2->Lines->Add(" // Save matrix values to txt file }");
  RichEdit2->Lines->Add(" StringList->SaveToFile(""ASCIIMtx.txt"");");
  RichEdit2->Lines->Add("} __finally {");
  RichEdit2->Lines->Add(" delete StringList;");
  RichEdit2->Lines->Add("}");
  RichEdit2->Lines->Add("");
  RichEdit2->Lines->Add("TStringList StringList = new TStringList();");
  RichEdit2->Lines->Add("Matrix tmpMtx;");
  RichEdit2->Lines->Add("try {");
  RichEdit2->SelAttributes->Style = RichEdit2->SelAttributes->Style << TFontStyle::fsItalic;
  RichEdit2->Lines->Add(" // get matrix values from text file }");
  RichEdit2->Lines->Add(" StringList->LoadFromFile(""ASCIIMtx.txt"");");
  RichEdit2->SelAttributes->Style = RichEdit2->SelAttributes->Style << TFontStyle::fsItalic;
  RichEdit2->Lines->Add(" // use tab = chr(9) as delimiter }");
  RichEdit2->Lines->Add(" tmpMtx->StringsToValues(StringList,#9);");
  RichEdit2->Lines->Add(" if (CheckBox1->Checked) ViewValues(tmpMtx);");
  RichEdit2->Lines->Add("} __finally {");
  RichEdit2->Lines->Add(" delete StringList;");
  RichEdit2->Lines->Add("}");

  AMtx.Size(20,20,true);
  AMtx.RandUniform(-1,2);
  TStringList *StringList = new TStringList();
  try
  {
	  AMtx.ValuesToStrings(StringList,kTab, ftaNone, "0.###;-0.###", "+0.###i;-0.###i");     // use tab = chr(9) as delimiter
	  StringList->SaveToFile("ASCIIMtx.txt");     // Save matrix values to txt file
  }
  __finally
  {
	  delete StringList;
  }

  StringList = new TStringList;
  try
  {
    // get matrix values from text file
    StringList->LoadFromFile("ASCIIMtx.txt");
	// use tab = chr(9) as delimiter
	tmpMtx.StringsToValues(StringList,kTab);
	if (CheckBox1->Checked) ViewValues(tmpMtx,"TEXT file",true,false,false);
  }
  __finally
  {
    delete StringList;
  }
}
//---------------------------------------------------------------------------

void __fastcall TFileH1::Button3Click(TObject *Sender)
{
    RichEdit2->Lines->Clear();

    RichEdit2->Lines->Add("Vector tmpVec;");
    RichEdit2->Lines->Add("AVec->Size(100,false);");
    RichEdit2->Lines->Add("AVec->RandGauss(0,1);");
    RichEdit2->Lines->Add("AVec->SortAscend();");
    RichEdit2->Lines->Add("TTable *Table1 = new TTable(this);");
    RichEdit2->Lines->Add("try");
    RichEdit2->Lines->Add("{");
    RichEdit2->Lines->Add("     Table1->DatabaseName = ""DBDemos"";");
    RichEdit2->Lines->Add("     Table1->TableType = ttDefault;");
    RichEdit2->Lines->Add("     Table1->TableName =""TableVec"";");
    RichEdit2->Lines->Add("     Table1->FieldDefs->Clear();");
    RichEdit2->Lines->Add("     TFieldDef *pNewFDef = Table1->FieldDefs->AddFieldDef();");
    RichEdit2->Lines->Add("     pNewFDef->Name = ""Field1"";");
    RichEdit2->Lines->Add("     pNewFDef->DataType = ftBlob;");
    RichEdit2->Lines->Add("     pNewFDef->Required = false;");
    RichEdit2->Lines->Add("     Table1->CreateTable();");
    RichEdit2->Lines->Add("     // save to table1");
    RichEdit2->Lines->Add("     Table1->Active = true;");
    RichEdit2->Lines->Add("     Table1->Insert();");
    RichEdit2->Lines->Add("     TStream *tmpBLOB = Table1->CreateBlobStream(Table1->FieldByName(""Field1""),bmReadWrite);");
    RichEdit2->Lines->Add("     try");
    RichEdit2->Lines->Add("     {");
    RichEdit2->Lines->Add("          AVec->SaveToStream(tmpBLOB, prDouble, rnTrunc);");
    RichEdit2->Lines->Add("          Table1->Post();");
    RichEdit2->Lines->Add("          tmpBLOB->Seek(0,0); // ""reset"" BLOB stream");
    RichEdit2->Lines->Add("          tmpVec->LoadFromStream(tmpBLOB);");
    RichEdit2->Lines->Add("          if (CheckBox1->Checked) ViewValues(tmpVec,""BLOB Stream"",true); ");
    RichEdit2->Lines->Add("     } ");
    RichEdit2->Lines->Add("     __finally ");
    RichEdit2->Lines->Add("     { ");
    RichEdit2->Lines->Add("          delete tmpBLOB;");
    RichEdit2->Lines->Add("     }");
    RichEdit2->Lines->Add("     }");
    RichEdit2->Lines->Add("__finally");
    RichEdit2->Lines->Add("{");
    RichEdit2->Lines->Add("     Table1->Active = false;");
    RichEdit2->Lines->Add("     Table1->DeleteTable();");
    RichEdit2->Lines->Add("     delete Table1;");
    RichEdit2->Lines->Add("}");

  #ifdef BDE

      Vector tmpVec;
      AVec->Size(100,false);
      AVec->RandGauss(0,1);
      AVec->SortAscend();
      TTable *Table1 = new TTable(this);
      try
      {
          Table1->DatabaseName = "DBDemos";
          Table1->TableType = ttDefault;
          Table1->TableName ="TableVec";
          Table1->FieldDefs->Clear();
          TFieldDef *pNewFDef = Table1->FieldDefs->AddFieldDef();
          pNewFDef->Name = "Field1";
          pNewFDef->DataType = ftBlob;
          pNewFDef->Required = false;
          Table1->CreateTable();
          // save to table1
          Table1->Active = true;
          Table1->Insert();
          TStream *tmpBLOB = Table1->CreateBlobStream(Table1->FieldByName("Field1"),bmReadWrite);
          try
          {
              AVec->SaveToStream(tmpBLOB, prDouble, rnTrunc);
              Table1->Post();
              tmpBLOB->Seek(0,0); // "reset" BLOB stream
              tmpVec->LoadFromStream(tmpBLOB);
              if (CheckBox1->Checked) ViewValues(tmpVec,"BLOB Stream",true); 
              }
          __finally
          {
              delete tmpBLOB;
          }
      }
      __finally");
      {");
      Table1->Active = false;");
      Table1->DeleteTable();");
      delete Table1;");
      }");

    #endif

}
//---------------------------------------------------------------------------

